File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.js
class Solution {
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) {
        let longest = 0;
        let left = 0;
        let maxf = 0;
        const freq = {};
        for (let right = 0; right < s.length; right++) {
            freq[s[right]] = s[right] in freq ? freq[s[right]] + 1 : 0;
            maxf = Math.max(maxf, freq[s[right]]);
            // window condition
            while (right - left - maxf > k) {
                freq[s[left]]--;
                left++;
            }
            longest = Math.max(right - left + 1, longest);
        }

        return longest;
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        right = left = longest = 0
        
        seen = set()

        while right < len(s):
            while s[right] in seen:
                seen.remove(s[left])
                left += 1
            seen.add(s[right])
            longest = max(longest, len(seen))
            right += 1
        
        return longest